### PR TITLE
SmartPointer derive-macro

### PR DESCRIPTION
**Pull request by dingxiangfei2009**
Possibly replacing #123472 for continued upkeep of the proposal rust-lang/rfcs#3621 and implementation of the tracking issue #123430.

cc @Darksonn @wedsonaf

### CODE REVIEW

**Comment by Nadrieril** [comment]
I don't feel competent reviewing this

r? compiler

**Comment by cjgillot** [comment]
r? compiler

File: compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                generics,
+                of_trait: Some(trait_ref),
+                self_ty: self_type.clone(),
+                items: ThinVec::new(),
+            })),
+        );
+        push(Annotatable::Item(item));
+    };
+
+    // Create unsized `self`, that is, one where the first type arg is replace with `__S`. For
```
Comment (davidtwco) [inline compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs:104]:
```suggestion
    // Create unsized `self`, that is, one where the pointee type arg is replaced with `__S`. For
```
Should it be this?

File: compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    // example, instead of `MyType<'a, T>`, it will be `MyType<'a, __S>`.
+    let s_ty = cx.ty_ident(span, Ident::new(sym::__S, span));
+    let mut alt_self_params = self_params;
+    alt_self_params[pointee_param_idx] = GenericArg::Type(s_ty.clone());
+    let alt_self_type = cx.ty_path(cx.path_all(span, false, vec![name_ident], alt_self_params));
+
+    // Find the first type parameter and add an `Unsize<__S>` bound to it.
```
Comment (davidtwco) [inline compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs:111]:
```suggestion
    // Find the pointee type parameter and add an `Unsize<__S>` bound to it.
```
Likewise

**Review by davidtwco — APPROVED** [review]
Obviously this derive is still undergoing lots of discussion in the RFC, but this seems like a decent start at an implementation that will allow experimentation and can evolve once the RFC settles.

r=me with comments addressed

File: compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs (hunk 1 above)
Comment (dingxiangfei2009) [reply compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs]: Yes!

Applied.

**Review by davidtwco — APPROVED** [review]

**Comment by cjgillot** [comment]
r? compiler

File: compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        );
+        push(Annotatable::Item(item));
+    };
+
+    // Create unsized `self`, that is, one where the `#[pointee]` type arg is replaced with `__S`. For
+    // example, instead of `MyType<'a, T>`, it will be `MyType<'a, __S>`.
+    let s_ty = cx.ty_ident(span, Ident::new(sym::__S, span));
+    let mut alt_self_params = self_params;
+    alt_self_params[pointee_param_idx] = GenericArg::Type(s_ty.clone());
+    let alt_self_type = cx.ty_path(cx.path_all(span, false, vec![name_ident], alt_self_params));
```
Comment (Kixunil) [inline compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs:109]: It looks like the trait bounds on pointee need to be copied for `__S` somewhere here. See #127647

**Pull request merged**